src/System.Private.CoreLib/shared/System/IO/Path.cs

```diff
@@ -444,6 +461,61 @@ public static string Join(string? path1, string? path2, string? path3)
             return Join(path1.AsSpan(), path2.AsSpan(), path3.AsSpan());
         }
 
+        public static string Join(string? path1, string? path2, string? path3, string? path4)
+        {
+            return Join(path1.AsSpan(), path2.AsSpan(), path3.AsSpan(), path4.AsSpan());
+        }
+
+        public static string Join(params string[] paths)
+        {
+            if (paths == null)
+            {
+                throw new ArgumentNullException(nameof(paths));
+            }
+
+            int finalSize = 0;
+            
+            for (int i = 0; i < paths.Length; i++)
+            {
+                if (paths[i] == null || paths[i].Length == 0)
+                {
+                    continue;
+                }
+
+                finalSize += paths[i].Length;
+
+                char ch = paths[i][paths[i].Length - 1];
+                if (!PathInternal.IsDirectorySeparator(ch))
```

Comment:
This isn't technically correct given that we check both sides with `Join`. Quite honestly I think it would be way better here to simply run through the set of paths tallying up the lengths. Something like:

``` C#
foreach (string path in paths) maxSize += path?.Length ?? 0;
maxSize += paths.Length - 1;

Span<char> stackBuffer = stackalloc char[260]; // Not strictly necessary, not sure how much it will help
var builder = new ValueStringBuilder(stackBuffer);
builder.EnsureCapacity(maxSize);
// loop through and append paths as below
return builder.ToString(); // This will return any rented buffer
```